modules/fedimint-dummy-client/src/lib.rs

```diff
@@ -1 +1,295 @@
+    ClientModule, DynPrimaryClientModule, IClientModule, PrimaryClientModule,
+};
+use fedimint_client::sm::{Context, ModuleNotifier, OperationId};
+use fedimint_client::transaction::{ClientInput, ClientOutput, TransactionBuilder};
+use fedimint_client::{Client, DynGlobalClientContext};
+use fedimint_core::api::GlobalFederationApi;
+use fedimint_core::core::{IntoDynInstance, KeyPair, ModuleInstanceId};
+use fedimint_core::db::{Database, ModuleDatabaseTransaction};
+use fedimint_core::module::{
+    CommonModuleGen, ExtendsCommonModuleGen, ModuleCommon, TransactionItemAmount,
+};
+use fedimint_core::{apply, async_trait_maybe_send, Amount, OutPoint};
 pub use fedimint_dummy_common as common;
+use fedimint_dummy_common::config::DummyClientConfig;
+use fedimint_dummy_common::{
+    DummyCommonGen, DummyInput, DummyModuleTypes, DummyOutput, DummyOutputOutcome,
+    DummyPrintMoneyRequest,
+};
+use secp256k1::{Secp256k1, XOnlyPublicKey};
+use states::DummyStateMachine;
+
+use crate::api::DummyFederationApi;
+use crate::db::DummyClientFundsKeyV0;
+
+pub mod api;
+mod db;
+mod states;
+
+/// Exposed API calls for client apps
+#[apply(async_trait_maybe_send!)]
+pub trait DummyClientExt {
+    /// Request the federation prints money for us
+    async fn print_money(&self, amount: Amount) -> anyhow::Result<()>;
+
+    /// Send money to another user
+    async fn send_money(&self, account: XOnlyPublicKey, amount: Amount)
+        -> anyhow::Result<OutPoint>;
+
+    /// Wait to receive money at an outpoint
+    async fn receive_money(&self, outpoint: OutPoint) -> anyhow::Result<()>;
+
+    /// The amount of funds we have
+    async fn total_funds(&self) -> Amount;
+
+    /// Return our account
+    fn account(&self) -> XOnlyPublicKey;
+}
+
+#[apply(async_trait_maybe_send!)]
+impl DummyClientExt for Client {
+    async fn print_money(&self, amount: Amount) -> anyhow::Result<()> {
+        let (id, dummy) = dummy_client(self);
+        let account = dummy.key.x_only_public_key().0;
+        let request = DummyPrintMoneyRequest { amount, account };
+        self.api().print_money(request).await?;
+        let funds = self.api().wait_for_money(account).await?;
+
+        // TODO: Not very nice to get to the module db
+        let mod_db = self.db().new_isolated(id);
+        let mut dbtx = mod_db.begin_transaction().await;
+        dbtx.insert_entry(&DummyClientFundsKeyV0, &funds).await;
+        dbtx.commit_tx().await;
+        Ok(())
+    }
+
+    async fn send_money(
+        &self,
+        account: XOnlyPublicKey,
+        amount: Amount,
+    ) -> anyhow::Result<OutPoint> {
+        let (id, dummy) = dummy_client(self);
+        let mut dbtx = self.db().begin_transaction().await;
+        let op_id = rand::random();
+
+        // TODO: Building a tx could be easier
+        // Create input using our own account
+        let input = dummy
+            .create_sufficient_input(&mut dbtx.with_module_prefix(id), op_id, amount)
+            .await?;
```

Comment:
Why are we making the assumption that the user wants to spend from the dummy module wallet? Idiomatically one would create an unfunded transaction here and leave it to `finalize_and_submit_transaction` to fund it using the primary module. That *might* be the dummy module if configured, but not necessarily.

Reply:
:+1: That could be useful for funding our ecash wallet...